src/map/status.c

```diff
@@ -12325,6 +12291,151 @@ bool status_readdb_scconfig(char* fields[], int columns, int current) {
 
 	return true;
 }
+
+/**
+ * Processes a refine_db.conf entry.
+ *
+ * @param *r	  Libconfig setting entry. It is expected to be valid and it
+ *                won't be freed (it is care of the caller to do so if
+ *                necessary)
+ * @param n       Ordinal number of the entry, to be displayed in case of
+ *                validation errors.
+ * @param *source Source of the entry (file name), to be displayed in case of
+ *                validation errors.
+ * @return # of the validated entry, or 0 in case of failure.
+ */
+int status_readdb_refine_libconfig_sub(config_setting_t *r, int n, const char *source) {
+	config_setting_t *t = NULL;
+	int i, type, bonus_per_level, random_bonus, random_bonus_start_level;
+	
+	if( !status->refinedb_lookup_const(r, "wLvl", &type) ) {
+		ShowWarning("status_readdb_refine_libconfig_sub: Invalid or missing wLvl in \"%s\", entry #%d, skipping.\n", source, n);
+		return 0;
+	} else if ( type < REFINE_TYPE_ARMOR || type >= REFINE_TYPE_MAX ) {
+		ShowWarning("status_readdb_refine_libconfig_sub: Invalid Type for entry #%d of \"%s\", skipping.\n", n, source);
+		return 0;
+	} else if( !status->refinedb_lookup_const(r, "StatsPerLevel", &bonus_per_level) ) {
+		ShowWarning("status_readdb_refine_libconfig_sub: Missing StatsPerLevel in entry #%d of \"%s\", skipping.\n", n, source);
+		return 0;
+	} else if( !status->refinedb_lookup_const(r, "RandomBonusStartLevel", &random_bonus_start_level) ) {
+		ShowWarning("status_readdb_refine_libconfig_sub: Missing RandomBonusStartLevel in entry #%d of \"%s\", skipping.\n", n, source);
+		return 0;
+	} else if ( !status->refinedb_lookup_const(r, "RandomBonusValue", &random_bonus) ) {
+		ShowWarning("status_readdb_refine_libconfig_sub: Missing RandomBonusValue in entry #%d of \"%s\", skipping.\n", n, source);
+		return 0;
+	}
+	
+	if ((t=libconfig->setting_get_member(r, "Rate")) && config_setting_is_list(t)) {
+		config_setting_t *tt = NULL;
+		bool duplicate[MAX_REFINE];
+		
+		memset(&duplicate,0,sizeof(duplicate));
+		
+		for(i=0; i < MAX_REFINE; i++) {
+			tt = libconfig->setting_get_elem(t, i);
+			
+			if (!tt)
```

Comment:
Could you explain the intent of this?  If an entry is missing, what happens? If we break out of the loop here (and the same way in the loop below), the server will report no errors